drivers/mbox/mbox_ipm.c

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2025 Dmitrii Sharshakov <d3dx12.xx@gmail.com>
+ *
+ * Based on mbox_nrfx_ipc.c which is:
+ * Copyright (c) 2021 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "zephyr/devicetree.h"
+#define DT_DRV_COMPAT zephyr_ipm_mbox
+
+#include <zephyr/drivers/mbox.h>
+#include <zephyr/drivers/ipm.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/irq.h>
+
+LOG_MODULE_REGISTER(mbox_ipm, CONFIG_MBOX_LOG_LEVEL);
+
+#define MBOX_IPM_CHANNELS DT_INST_PROP(0, channels)
+
+struct mbox_ipm_data {
+	mbox_callback_t cb[MBOX_IPM_CHANNELS];
+	void *user_data[MBOX_IPM_CHANNELS];
+	const struct device *dev;
+	uint32_t enabled_mask;
+};
+
+static struct mbox_ipm_data ipm_mbox_data;
+
+static struct mbox_ipm_conf {
+	const struct device *ipm_dev;
+} ipm_mbox_conf = {
+	.ipm_dev = DEVICE_DT_GET(DT_PARENT(DT_INST(0, zephyr_ipm_mbox))),
```

Comment:
Another option would be a phandle passed as a DT property. I chose parent to make sure DT infrastructure ensures correct init order.